receiver/awslambdareceiver/internal/s3.go

```diff
@@ -0,0 +1,106 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awslambdareceiver/internal"
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+)
+
+// s3API abstracts out the S3 APIs allowing mocking for tests
+type s3API interface {
+	GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+	ListObjectsV2(ctx context.Context, params *s3.ListObjectsV2Input, optFns ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
+	DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error)
+}
+
+// S3Service define services exposed for consumers
+type S3Service interface {
+	ReadObject(ctx context.Context, bucketName, objectKey string) ([]byte, error)
+	ListObjects(ctx context.Context, bucketName, continuationToken, prefix string) (*s3.ListObjectsV2Output, error)
+	DeleteObject(ctx context.Context, bucketName, objectKey string) error
+}
+
+// S3Provider expose contract to get S3Service
+type S3Provider interface {
+	GetService(ctx context.Context) (S3Service, error)
+}
+
+// S3ServiceProvider is a helper to derive reusable, lazy initialized S3Service for any dependent component.
+type S3ServiceProvider struct {
+	service S3Service
+	once    sync.Once
+	error   error
+}
+
+func (p *S3ServiceProvider) GetService(ctx context.Context) (S3Service, error) {
+	p.once.Do(func() {
```

Comment:
Why are we using a sync.Once here? It seems unnecessary, and complicates use of ctx.

Reply:
Good catch! GetService() is only called during initialization, not per-invocation, so 
sync.Once added no value. I have simplified the code now.